### ISSUE TITLE
Add support for Amazon CodeCommit

### ISSUE THREAD

**Issue by pditommaso** [nextflow-io/nextflow#61]
[CodeCommit](http://aws.amazon.com/codecommit/) is a code repository hosted on Amazon AWS. 

The main benefit is that it allows large files to be stored on it (up to 2GB per file).

Configuration doc is available at [this link](http://docs.aws.amazon.com/codecommit/latest/userguide/setting-up-topnode.html).

API documentation is available at [this link](http://docs.aws.amazon.com/codecommit/latest/APIReference/Welcome.html).

**Comment by pditommaso** [comment]
At this time this feature is not relevant. Closing for now.

**Comment by wleepang** [comment]
Curious if there is interest in re-opening this issue and targeting accessing AWS CodeCommit using `git-remote-codecommit` over HTTPS as documented here:

https://docs.aws.amazon.com/codecommit/latest/userguide/setting-up-git-remote-codecommit.html

This would allow users to run workflows in a privately hosted AWS CodeCommit repo via:

```
nextflow run codecommit::<region>://<repository-name> ...
```

**Comment by pditommaso** [comment]
Don't remember exactly which technical issue I've found adding the support for CodeCommit. Does it support HTTP/S transport?

**Comment by wleepang** [comment]
Yes, CodeCommit supports both SSH and HTTPS transport.  For HTTPS, there are a couple options:

* Using git-remote-codecommit: https://docs.aws.amazon.com/codecommit/latest/userguide/setting-up-git-remote-codecommit.html
* Using a credential helper: https://docs.aws.amazon.com/codecommit/latest/userguide/setting-up-https-unixes.html

**Comment by pditommaso** [comment]
A pull request is welcome for this feature. The GitHub implementation can be used as a reference for it https://github.com/nextflow-io/nextflow/blob/b49e6d74dae782def6be2b087d06fdd8fa6b9291/modules/nextflow/src/main/groovy/nextflow/scm/GithubRepositoryProvider.groovy

**Comment by jcurado-flomics** [comment]
+1! I was also just looking for this. 
It would be great to have it

**Comment by wleepang** [comment]
@pditommaso - looking through the base provider, and wanted to clarify - does Nextflow primarily use HTTP/S API calls to get repo files?  Could there be an option to rely on an installed version of `git`?

**Comment by pditommaso** [comment]
NF uses the provider REST API to fetch the clone URL and read the pipeline config file(s) remotely i.e. without cloning it. 

Once it fetched the clone url, it pulls the project using the embedded Git client: 

https://github.com/nextflow-io/nextflow/blob/55be0762950184114877519fdb810afc199478f4/modules/nextflow/src/main/groovy/nextflow/scm/AssetManager.groovy#L571

Therefore don't think an installed version of `git` would help a lot. Not sure CodeCommit has a REST API to fetch repository metadata. If I remember well this was the problem I met when trying to implement the support for it.

**Comment by wleepang** [comment]
While you can you use a REST API with CodeCommit, you would need to handle the Sigv4 signing process to access it. These actions are encapsulated by the AWS SDK. So actions to get file contents and read branches would be SDK calls.

**Comment by pditommaso** [comment]
Therefore it needs to be authenticated via the AWS SDK for Java, that looks feasible.

**Comment by wleepang** [comment]
Is there somewhere else besides `scm/ProviderConfig.groovy` that needs to have the provider registered? I've got the AWS CodeCommit provider coded up, but the build doesn't recognize it. Instead, I get the following:

```
Unknown repository provider: `codecommit`'. Did you mean?
  github
  gitlab
  gitea
  bitbucket
```

**Comment by wleepang** [comment]
In `scm/ProviderConfig.groovy` I have the following:

```groovy
    static private void addDefaults(List<ProviderConfig> result) {
        if( !result.find{ it.name == 'github' })
            result << new ProviderConfig('github')

        if( !result.find{ it.name == 'gitlab' })
            result << new ProviderConfig('gitlab')

        if( !result.find{ it.name == 'gitea' })
            result << new ProviderConfig('gitea')

        if( !result.find{ it.name == 'bitbucket' })
            result << new ProviderConfig('bitbucket')
        
        if( !result.find{ it.name == 'codecommit' })
            result << new ProviderConfig('codecommit')
    }
```

**Comment by wleepang** [comment]
I figured it out - didn't see the instructions to run `launch.sh` after `make compile`.

**Comment by pditommaso** [comment]
Added a tentative implementation [here](https://github.com/nextflow-io/nextflow/pull/2794). (saw only now the `WIP`, hope there isn't too much overlap)

**Comment by pditommaso** [comment]
Solved by 80fba6e9 and 296d7add
